File: Sprint6/C/C++/main.cpp
// https://contest.yandex.ru/contest/25069/problems/C/
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void dfs(int v, const vector<vector<int>>& g, vector<bool>& visited) {
    visited[v] = true;
    cout << v + 1 << " ";

    for (int neighbor : g[v]) {
        if (!visited[neighbor]) {
            dfs(neighbor, g, visited);
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<vector<int>> g(n);
    vector<bool> visited(n, false);

    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        --u;
        --v;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    for (auto& neighbors : g) {
        sort(neighbors.begin(), neighbors.end());
    }

    int start_v;
    cin >> start_v;
    --start_v;

    dfs(start_v, g, visited);

    cout << endl;
    return 0;
}

File: Sprint6/C/Rust/src/main.rs
// https://contest.yandex.ru/contest/25069/problems/C/

use std::io::{self, BufRead};

fn dfs(v: usize, g: &Vec<Vec<usize>>, visited: &mut Vec<bool>) {
    visited[v] = true;
    print!("{} ", v + 1);

    for &neighbor in &g[v] {
        if !visited[neighbor] {
            dfs(neighbor, g, visited);
        }
    }
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let first_line = lines.next().unwrap().unwrap();
    let nm: Vec<usize> = first_line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    let (n, m) = (nm[0], nm[1]);

    let mut g: Vec<Vec<usize>> = vec![vec![]; n];
    let mut visited: Vec<bool> = vec![false; n];

    for _ in 0..m {
        let edge: Vec<usize> = lines
            .next()
            .unwrap()
            .unwrap()
            .split_whitespace()
            .map(|x| x.parse::<usize>().unwrap() - 1)
            .collect();
        let (u, v) = (edge[0], edge[1]);
        g[u].push(v);
        g[v].push(u);
    }

    for neighbors in &mut g {
        neighbors.sort();
    }

    let start_v: usize = lines.next().unwrap().unwrap().parse::<usize>().unwrap() - 1;

    dfs(start_v, &g, &mut visited);

    println!();
    Ok(())
}
